### PR TITLE
Extract ``GPUEngine`` config options at translation time

### PR DESCRIPTION
**Pull request by rjzamora**
## Description
Follow up to https://github.com/rapidsai/cudf/pull/16944
That PR added `config: GPUEngine` to the arguments of every `IR.do_evaluate` function. In order to simplify future multi-GPU development, this PR extracts the necessary configuration argument at `IR` translation time instead.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf_polars/cudf_polars/dsl/translate.py [hunk 1]
```diff
@@ -39,10 +41,13 @@ class Translator:
     ----------
     visitor
         Polars NodeTraverser object
+    config
+        GPU engine configuration.
     """
 
-    def __init__(self, visitor: NodeTraverser):
+    def __init__(self, visitor: NodeTraverser, config: GPUEngine):
```
Comment (rjzamora) [inline python/cudf_polars/cudf_polars/dsl/translate.py:48]: Wasn't sure if we want `config` to be a required argument. I suppose we could also fall back to `pl.GPUEngine()` if nothing is passed in (would simplify any testing and debugging cases where a special configuration isn't needed).

Comment (brandon-b-miller) [reply python/cudf_polars/cudf_polars/dsl/translate.py:48]: My intuition is that it should be explicit as we have here already. I feel like instantiating configs by default too far into our internals might make it harder to figure out where the config being observed is came from in the first place.

**Review by brandon-b-miller — APPROVED** [review]
I'm +1 here.

File: python/cudf_polars/cudf_polars/dsl/ir.py [hunk 2]
```diff
@@ -321,6 +319,7 @@ def __init__(
         self.typ = typ
         self.reader_options = reader_options
         self.cloud_options = cloud_options
+        self.config_options = config_options
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply python/cudf_polars/cudf_polars/dsl/ir.py:322]: Ah, good catch - Added.

File: python/cudf_polars/cudf_polars/testing/asserts.py [hunk 3]
```diff
@@ -122,7 +122,7 @@ def assert_ir_translation_raises(q: pl.LazyFrame, *exceptions: type[Exception])
     AssertionError
        If the specified exceptions were not raised.
     """
-    translator = Translator(q._ldf.visit())
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply python/cudf_polars/cudf_polars/testing/asserts.py:125]: I cant remember "why" I included `raise_on_fail=True` here, but it doesn't seem to impact any of the tests (So, I just removed it.).

File: python/cudf_polars/cudf_polars/dsl/translate.py [hunk 4]
```diff
@@ -223,11 +228,17 @@ def _(
         skip_rows, n_rows = n_rows
 
     row_index = file_options.row_index
+    config_options = (
+        {"parquet_options": translator.config.config.get("parquet_options", {})}
+        if typ == "parquet"
+        else {}
+    )
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply python/cudf_polars/cudf_polars/dsl/translate.py]: Cool - Works for me.

**Comment by rjzamora** [comment]
@wence- - I think this is ready for a final review. Let me know if we are still good to target 24.12.

**Review by wence- — APPROVED** [review]

**Comment by vyasr** [comment]
The error code in the polars tests here is the same that we observed in a few runs on https://github.com/rapidsai/cudf/pull/17362 such as [this one](https://github.com/rapidsai/cudf/actions/runs/11918840035/job/33233279030). @wence- have you seen that anywhere else? Is it something that we need to do any debugging of?

**Pull request merged**